File: tiktaktoe/controllers/tictactoe.py
from odoo import http
from odoo.http import request

class TicTacToeController(http.Controller):
    
    @http.route('/tic_tac_toe/move', type='json', auth="user")
    def make_move(self, game_id, position, player):
        game = request.env['tic.tac.toe'].browse(game_id)
        if game.state != 'ongoing':
            return {'error': 'Game has already ended.'}

        board = list(game.board)
        if board[position] != '-':
            return {'error': 'Invalid move, position already taken.'}

        board[position] = player
        game.board = ''.join(board)

        winner = self.check_winner(board)
        if winner:
            game.winner = winner
            game.state = 'finished'

        return {'board': game.board, 'winner': game.winner}

    def check_winner(self, board):
        win_patterns = [(0,1,2), (3,4,5), (6,7,8), (0,3,6), (1,4,7), (2,5,8), (0,4,8), (2,4,6)]
        for a, b, c in win_patterns:
            if board[a] == board[b] == board[c] and board[a] != '-':
                return board[a]
        if '-' not in board:
            return 'draw'
        return None

File: tiktaktoe/static/src/js/tictactoe.js
odoo.define('tic_tac_toe.tictactoe', function (require) {
    "use strict";

    var core = require('web.core');
    var rpc = require('web.rpc');

    $(document).on('click', '.tictactoe-cell', function () {
        var cell = $(this);
        var game_id = cell.data('game-id');
        var position = cell.data('position');
        var player = $('#current-player').val();

        rpc.query({
            route: '/tic_tac_toe/move',
            params: { game_id: game_id, position: position, player: player }
        }).then(function (result) {
            if (result.error) {
                alert(result.error);
            } else {
                cell.text(player);
                if (result.winner) {
                    alert("Winner: " + result.winner);
                }
            }
        });
    });
});
